### ISSUE TITLE
Train on CPU?

### ISSUE THREAD

**Issue by vlavorini** [ultralytics/ultralytics#1104]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

What options should I use to run the training by using only the CPU?

### Additional

I tried by reducing the batch size down to 1, but the kernel  keeps dying.

I'm using Jupyuter lab, python 3.9.12, yolo v8n

**Comment by binn77** [comment]
> ### Search before asking
> * [x]  I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.
> 
> ### Question
> What options should I use to run the training by using only the CPU?
> 
> ### Additional
> I tried by reducing the batch size down to 1, but the kernel keeps dying.
> 
> I'm using Jupyuter lab, python 3.9.12, yolo v8n

`yolo classify train model=yolov8s-cls.pt data=C:/temp/test/  device=cpu`

**Comment by vlavorini** [comment]
Thank you, but even with such option  the kernel dies.

For clarity, I'm trying to perform object detection with the following code:

    from ultralytics import YOLO

    # Load a model
    model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)

    # Train the model
    model.train(data="./products.yaml", epochs=3, imgsz=640, batch=1, device='cpu')

The output (before crashing) is:

Ultralytics YOLOv8.0.43  Python-3.9.12 torch-1.13.1 CPU
yolo\engine\trainer: task=detect, mode=train, model=yolov8n.pt, data=./products.yaml, epochs=3, patience=50, batch=1, imgsz=640, save=True, save_period=-1, cache=False, device=cpu, workers=8, project=None, name=None, exist_ok=False, pretrained=False, optimizer=SGD, verbose=True, seed=0, deterministic=True, single_cls=False, image_weights=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, min_memory=False, overlap_mask=True, mask_ratio=4, dropout=False, val=True, split=val, save_json=False, save_hybrid=False, conf=0.001, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=ultralytics/assets/, show=False, save_txt=False, save_conf=False, save_crop=False, hide_labels=False, hide_conf=False, vid_stride=1, line_thickness=3, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.001, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, fl_gamma=0.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, cfg=None, v5loader=False, tracker=botsort.yaml, save_dir=runs\detect\train15
Overriding model.yaml nc=80 with nc=4

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.Conv                  [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.Conv                  [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.C2f                   [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.Conv                  [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.C2f                   [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.Conv                  [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.C2f                   [128, 128, 2, True]           
  7                  -1  1    295424  ultralytics.nn.modules.Conv                  [128, 256, 3, 2]              
  8                  -1  1    460288  ultralytics.nn.modules.C2f                   [256, 256, 1, True]           
  9                  -1  1    164608  ultralytics.nn.modules.SPPF                  [256, 256, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.Concat                [1]                           
 12                  -1  1    148224  ultralytics.nn.modules.C2f                   [384, 128, 1]                 
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.Concat                [1]                           
 15                  -1  1     37248  ultralytics.nn.modules.C2f                   [192, 64, 1]                  
 16                  -1  1     36992  ultralytics.nn.modules.Conv                  [64, 64, 3, 2]                
 17            [-1, 12]  1         0  ultralytics.nn.modules.Concat                [1]                           
 18                  -1  1    123648  ultralytics.nn.modules.C2f                   [192, 128, 1]                 
 19                  -1  1    147712  ultralytics.nn.modules.Conv                  [128, 128, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.Concat                [1]                           
 21                  -1  1    493056  ultralytics.nn.modules.C2f                   [384, 256, 1]                 
 22        [15, 18, 21]  1    752092  ultralytics.nn.modules.Detect                [4, [64, 128, 256]]           
Model summary: 225 layers, 3011628 parameters, 3011612 gradients, 8.2 GFLOPs

Transferred 319/355 items from pretrained weights
optimizer: SGD(lr=0.01) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.001), 63 bias
train: Scanning C:\Users\vincenzo.lavorini\Devel\hackaton\datasets\train\labels.cache... 2908 images, 0 backgrounds, 0 corrupt: 100%|██████████| 2908/2908 [00:00<?, ?it/s]
val: Scanning C:\Users\vincenzo.lavorini\Devel\hackaton\datasets\val\labels.cache... 726 images, 0 backgrounds, 0 corrupt: 100%|██████████| 726/726 [00:00<?, ?it/s]
Plotting labels to runs\detect\train15\labels.jpg...

**Comment by binn77** [comment]
> Thank you, but even with such option the kernel dies.
> 
> For clarity, I'm trying to perform object detection with the following code:
> 
> ```
> from ultralytics import YOLO
> 
> # Load a model
> model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)
> 
> # Train the model
> model.train(data="./products.yaml", epochs=3, imgsz=640, batch=1, device='cpu')
> ```
> 
> The output (before crashing) is:
> 
> Ultralytics YOLOv8.0.43 Python-3.9.12 torch-1.13.1 CPU yolo\engine\trainer: task=detect, mode=train, model=yolov8n.pt, data=./products.yaml, epochs=3, patience=50, batch=1, imgsz=640, save=True, save_period=-1, cache=False, device=cpu, workers=8, project=None, name=None, exist_ok=False, pretrained=False, optimizer=SGD, verbose=True, seed=0, deterministic=True, single_cls=False, image_weights=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, min_memory=False, overlap_mask=True, mask_ratio=4, dropout=False, val=True, split=val, save_json=False, save_hybrid=False, conf=0.001, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=ultralytics/assets/, show=False, save_txt=False, save_conf=False, save_crop=False, hide_labels=False, hide_conf=False, vid_stride=1, line_thickness=3, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.001, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, fl_gamma=0.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, cfg=None, v5loader=False, tracker=botsort.yaml, save_dir=runs\detect\train15 Overriding model.yaml nc=80 with nc=4
> 
> ```
>                from  n    params  module                                       arguments                     
> ```
> 
> 0 -1 1 464 ultralytics.nn.modules.Conv [3, 16, 3, 2] 1 -1 1 4672 ultralytics.nn.modules.Conv [16, 32, 3, 2] 2 -1 1 7360 ultralytics.nn.modules.C2f [32, 32, 1, True] 3 -1 1 18560 ultralytics.nn.modules.Conv [32, 64, 3, 2] 4 -1 2 49664 ultralytics.nn.modules.C2f [64, 64, 2, True] 5 -1 1 73984 ultralytics.nn.modules.Conv [64, 128, 3, 2] 6 -1 2 197632 ultralytics.nn.modules.C2f [128, 128, 2, True] 7 -1 1 295424 ultralytics.nn.modules.Conv [128, 256, 3, 2] 8 -1 1 460288 ultralytics.nn.modules.C2f [256, 256, 1, True] 9 -1 1 164608 ultralytics.nn.modules.SPPF [256, 256, 5] 10 -1 1 0 torch.nn.modules.upsampling.Upsample [None, 2, 'nearest'] 11 [-1, 6] 1 0 ultralytics.nn.modules.Concat [1] 12 -1 1 148224 ultralytics.nn.modules.C2f [384, 128, 1] 13 -1 1 0 torch.nn.modules.upsampling.Upsample [None, 2, 'nearest'] 14 [-1, 4] 1 0 ultralytics.nn.modules.Concat [1] 15 -1 1 37248 ultralytics.nn.modules.C2f [192, 64, 1] 16 -1 1 36992 ultralytics.nn.modules.Conv [64, 64, 3, 2] 17 [-1, 12] 1 0 ultralytics.nn.modules.Concat [1] 18 -1 1 123648 ultralytics.nn.modules.C2f [192, 128, 1] 19 -1 1 147712 ultralytics.nn.modules.Conv [128, 128, 3, 2] 20 [-1, 9] 1 0 ultralytics.nn.modules.Concat [1] 21 -1 1 493056 ultralytics.nn.modules.C2f [384, 256, 1] 22 [15, 18, 21] 1 752092 ultralytics.nn.modules.Detect [4, [64, 128, 256]] Model summary: 225 layers, 3011628 parameters, 3011612 gradients, 8.2 GFLOPs
> 
> Transferred 319/355 items from pretrained weights optimizer: SGD(lr=0.01) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.001), 63 bias train: Scanning C:\Users\vincenzo.lavorini\Devel\hackaton\datasets\train\labels.cache... 2908 images, 0 backgrounds, 0 corrupt: 100%|██████████| 2908/2908 [00:00<?, ?it/s] val: Scanning C:\Users\vincenzo.lavorini\Devel\hackaton\datasets\val\labels.cache... 726 images, 0 backgrounds, 0 corrupt: 100%|██████████| 726/726 [00:00<?, ?it/s] Plotting labels to runs\detect\train15\labels.jpg...

Try to reduce this value to workers=2

**Comment by vlavorini** [comment]
Also with 2 workers crashes. Note that the system has 16 GB of RAM, CPU i5 (one years old PC)

**Comment by binn77** [comment]
> Also with 2 workers crashes. Note that the system has 16 GB of RAM, CPU i5 (one years old PC)

There may be a problem with images for training. Try to check, for example, to create 2 classes and in each 10 images, see if there will be departures.

**Comment by vlavorini** [comment]
I tried with Linux, and everything works. So there was probably some issue with libraries.

Thank you anyway for the hints

**Comment by pderrenger** [comment]
@vlavorini you're welcome! I'm glad to hear that things are working in Linux. If you encounter any further issues or have additional questions, feel free to ask. Good luck with your project!
